### PR TITLE
Adds support for multi vector audience definitions.

### PR DESCRIPTION
**Pull request by naripok**
I squashed the commits, so the PR is kinda big...
It implements the proposed changes an refactor the translation and createCondition logic.
Also, use enum type for filter options. If it breaks existing logic, I can rollback it 
and rewrite with a union type.

Cheers!

### CODE REVIEW

File: docs/audiences.md [hunk 1]
```diff
@@ -33,7 +33,7 @@ export const exampleAudience: AudienceDefinition = {
     // The query property to look up, this is the name of the key that will be looked up in the stored page view features object
     queryProperty: 'keywords',
     // The name of the function to use for filtering the page view features
-    queryFilterComparisonType: 'arrayIntersects',
+    queryFilterComparisonType: QueryFilterComparisonType.ARRAY_INTERSECTS,
```
Comment (ydennisy) [inline docs/audiences.md:36]: nice!

File: test/audienceCosineSimMultipleVectors.test.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
+import { AudienceDefinition, QueryFilterComparisonType } from '../types';
+import { edkt } from '../src';
+
+/* TODO Mock localStorage and merge single and
```
Comment (ydennisy) [inline test/audienceCosineSimMultipleVectors.test.ts:4]: What do you mean here by mock localStorage?

File: test/cosineSimilarityEngineCondition.test.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
+import { check } from '../src/engine';
+import {
+  EngineCondition,
+  QueryFilterComparisonType,
+  CosineSimilarityFilter,
+  PageView
+} from '../types';
+
+/* TODO queryProperty based tests for engine behaviors
```
Comment (ydennisy) [inline test/cosineSimilarityEngineCondition.test.ts:9]: Could you please open a new issue to discuss this TODO?

File: test/vectorDistanceEngineCondition.test.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        name: 'ge',
+        args: 1,
+      },
+    },
+  ],
+};
+
+describe('VectorDistance condition test', () => {
+  describe('Vector condition', () => {
+    it('matches the page view if similar enough', async () => {
```
Comment (ydennisy) [inline test/vectorDistanceEngineCondition.test.ts:71]: Can we word this a little more clearly, similar enough -> above threshold

File: types/index.ts [hunk 5]
```diff
@@ -54,82 +54,71 @@ export interface MatchedAudience {
...
+  queryValue: VectorQueryValue[];
+  queryFilterComparisonType: QueryFilterComparisonType.COSINE_SIMILARITY;
+}
+
+export type AudienceDefinitionFilter =
+  | VectorDistanceFilter
+  | CosineSimilarityFilter
+  | ArrayIntersectsFilter
+
+// Not a good name...
```
Comment (ydennisy) [inline types/index.ts:102]: Yes I agree, we cannot merge this like this :)

File: src/engine/condition.ts [hunk 6]
```diff
@@ -1,45 +1,51 @@
...
-    .flat();
+  .map((query) => {
+    return pageViews.filter((pageView) => {
+      const queryFeatures: PageFeatureResult = pageView.features[query.queryProperty];
+      return isArrayIntersectsFilter(query)
+        ? isArrayIntersects(queryFeatures, query)
+        : isVectorDistanceFilter(query)
+        ? isVectorDistanceLesserThanThreshold(queryFeatures, query)
+        : isCosineSimilarityFilter(query)
+        ? isCosineSimilarityLesserThanThreshold(queryFeatures, query)
```
Comment (ydennisy) [inline src/engine/condition.ts:35]: I am not a fan of the nested ternary, it looks neat but its hard to follow, I am happy to leave it for now if you do not have any other ideas on how to handle this neatly. Quick one would be:

```
const filterType = checkQueryFilter(query);
filter(queryFeatures, query, filterType);
```
But I guess this could have issues with TS...

Also `isVectorDistanceLesserThanThreshold`, we match on greater than threshold.

File: src/engine/condition.ts [hunk 7]
```diff
@@ -1,45 +1,51 @@
...
+        : true;  /* TODO is this right?
```
Comment (ydennisy) [inline src/engine/condition.ts:36]: No this is wrong - I cannot think of why this would be here.

If none of the filters work we should not match by default.

Good spot 👍

File: src/engine/condition.ts [hunk 8]
```diff
@@ -56,3 +62,63 @@ const createCondition = (condition: EngineCondition) => (
 };
 
 export default createCondition;
+
+/* Matching conditions pieces */
+
+const isArrayIntersects =
```
Comment (ydennisy) [inline src/engine/condition.ts:68]: I think this is getting quite confusing, we have `filters` code:
- `src/engine/filters`
- `src/engine/condition`
- `src/utils`

Can we consolidate and name so it is easy to reason about how they interact?

Makes sense to have `src/engine/filters` export just a single filter for use elsewhere. 

WDYT?

File: src/engine/condition.ts [hunk 9]
```diff
@@ -56,3 +62,63 @@ const createCondition = (condition: EngineCondition) => (
...
+  (features: PageFeatureResult, query: EngineConditionQuery<ArrayIntersectsFilter>): boolean =>
+  !!features &&
+  versionMatches(features, query) &&
+  stringArrayFilterMatches(filters.arrayIntersects, features, query);
+
+const isVectorDistanceLesserThanThreshold =
```
Comment (ydennisy) [inline src/engine/condition.ts:74]: I am surprised lint has not picked up on this... but this should be declared prior to use.

File: src/utils/index.ts [hunk 10]
```diff
@@ -63,3 +71,21 @@ export const isVectorQueryValue = (
   value: any
 ): value is VectorQueryValue =>
   isNumberArray(value.vector) && typeof value.threshold === 'number';
+
+export const isArrayIntersectsFilter = (
```
Comment (ydennisy) [inline src/utils/index.ts:75]: See comment on consolidating math + filters.

File: src/engine/translate.ts [hunk 11]
```diff
@@ -1,49 +1,37 @@
...
+  AudienceDefinition,
+  EngineCondition,
+  EngineConditionQuery,
+  AudienceDefinitionFilter,
+} from '../../types';
 
+/*
+ * I'm maintaining the union over the translation layer
+ * so the compiler can discriminate it further bellow in the computation
+ * TODO Better audienceDefinition validation
```
Comment (ydennisy) [inline src/engine/translate.ts:11]: What do you have in mind?

**Review by ydennisy — CHANGES_REQUESTED** [review]

File: src/engine/condition.ts (hunk 9 above)
Comment (naripok) [reply src/engine/condition.ts:74]: I guess `no-use-before-define` is not enabled by the [recommended settings](https://eslint.org/docs/rules/).

File: src/engine/condition.ts (hunk 8 above)
Comment (naripok) [reply src/engine/condition.ts:68]: I agree that it is getting confusing. The naming is bad too.

I'll study the idea of how to simplify the filter interface.

File: src/engine/condition.ts (hunk 6 above)
Comment (naripok) [reply src/engine/condition.ts:35]: I agree. This is used to satisfy TS checker on the type specialization.

On your idea, I think, `checkQueryFilter` return type would be yet an union and so the `filter` function would not have been specialized.

I'll try to come up with a better solution.

File: test/cosineSimilarityEngineCondition.test.ts (hunk 3 above)
Comment (naripok) [reply test/cosineSimilarityEngineCondition.test.ts:9]: This should have been `property based tests for engine behaviors`.

I'm going to write a minimal example to discuss this topic further.

File: src/engine/translate.ts (hunk 11 above)
Comment (naripok) [reply src/engine/translate.ts:11]: This transformation is actually an identity over its attributes. Being so, additional validation here is likely not needed given the params were validated in the above layers.

Actually, this function is not exactly a translation, or an adapter. This is a union between the audience definition attributes and the rule attributes. The rule ones are just constants for the moment, but I assume they will not remain this way much longer, right?

File: test/audienceCosineSimMultipleVectors.test.ts (hunk 2 above)
Comment (naripok) [reply test/audienceCosineSimMultipleVectors.test.ts]: In the current setting, localStorage as well as ViewStore, MatchedAudienceStore and CachedAudienceStore, are being used raw and the asserting are being made over their contents on each test suite.
This setting is weak as it cannot assure that their methods are being called with the proper expected inputs.
It can happens, for an example, that the actual content of one of these was inserted on some unrelated test suite and is being
used spuriously on an assertion.
This setting also does not work well on a property based setup where the methods should get called a lot of times with random inputs.

I'll try to improve on this tomorrow when working on the property based implementations.

Comment (ydennisy) [reply test/audienceCosineSimMultipleVectors.test.ts]: gotha! sounds good.

File: src/engine/filters/index.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  arrayIntersectsCondition,
+  vectorDistanceCondition,
+  cosineSimilarityCondition,
+} from './conditions';
+
+export const matchesQuery = (
+  query: EngineConditionQuery<AudienceDefinitionFilter>,
+  queryFeatures: PageFeatureResult
+): boolean => {
+  // Can queryFeatures be empty?
```
Comment (ydennisy) [inline src/engine/filters/index.ts:17]: Do you mean can the page views have empty features, yes they can - but in that case they should just be skipped.

File: src/engine/filters/index.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+  if (!queryFeatures || !versionMatches(queryFeatures, query)) {
+    return false;
+  }
+
+  // matches if any of the conditions are satisfied
+  return [
+    arrayIntersectsCondition,
+    vectorDistanceCondition,
+    cosineSimilarityCondition,
+  ].some((match) => match(queryFeatures, query));
```
Comment (ydennisy) [inline src/engine/filters/index.ts:27]: This feels a little inefficient to just run all the conditions on all the page views?

File: src/engine/filters/index.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+import {
+  versionMatches,
+  arrayIntersectsCondition,
+  vectorDistanceCondition,
+  cosineSimilarityCondition,
+} from './conditions';
+
+export const matchesQuery = (
+  query: EngineConditionQuery<AudienceDefinitionFilter>,
+  queryFeatures: PageFeatureResult
```
Comment (ydennisy) [inline src/engine/filters/index.ts:15]: I think the naming here can be changed, `query` and `queryFeatures` sounds like they come from the same object i.e `queryFeature` is a subset of `query`.

PageViewFeatures or PageFeatures I feel is a better way to describe what is stored client side based on user page visits/

File: src/engine/translate.ts [hunk 15]
```diff
@@ -1,62 +1,51 @@
...
+  EngineCondition,
+  EngineConditionQuery,
+  AudienceDefinitionFilter,
+} from '../../types';
 
+/*
+ * Audience to Engine adapter
+ *
+ * Maintains the union over the translation layer
+ * so the type can discriminated further bellow
```
Comment (ydennisy) [inline src/engine/translate.ts:12]: typo: below

File: src/engine/translate.ts [hunk 16]
```diff
@@ -1,62 +1,51 @@
...
-import { isStringArray, isVectorQueryValue } from '../utils';
+import {
+  Audience,
+  EngineCondition,
+  EngineConditionQuery,
+  AudienceDefinitionFilter,
+} from '../../types';
 
+/*
+ * Audience to Engine adapter
```
Comment (ydennisy) [inline src/engine/translate.ts:9]: adapter -> translation

File: types/index.ts [hunk 17]
```diff
@@ -54,82 +54,70 @@ export interface MatchedAudience {
...
+export type AudienceDefinition = {
+  featureVersion: number;
+  ttl: number;
+  lookBack: number;
+  occurrences: number;
+  queryProperty: string;
+} & AudienceDefinitionFilter;
+
+export interface Audience {
+  accountId?: Record<string, AudienceState>;
```
Comment (ydennisy) [inline types/index.ts:111]: I think this can be removed - is it used anywhere?

This is an internal concept, not needed in the open source lib.

File: types/index.ts [hunk 18]
```diff
@@ -54,82 +54,70 @@ export interface MatchedAudience {
...
+
+export type AudienceDefinition = {
+  featureVersion: number;
+  ttl: number;
+  lookBack: number;
+  occurrences: number;
+  queryProperty: string;
+} & AudienceDefinitionFilter;
+
+export interface Audience {
```
Comment (ydennisy) [inline types/index.ts:110]: `AudienceDefinition` && `AudienceQueryDefinition` ?

I think just Audience is a little loose?

WDYT?

File: types/index.ts [hunk 19]
```diff
@@ -54,82 +54,70 @@ export interface MatchedAudience {
...
+  accountId?: Record<string, AudienceState>;
   id: string;
   version: number;
+  name: string;
```
Comment (ydennisy) [inline types/index.ts:114]: This should be optional.

**Review by ydennisy — CHANGES_REQUESTED** [review]

File: src/engine/filters/index.ts (hunk 14 above)
Comment (naripok) [reply src/engine/filters/index.ts:15]: I agree

File: types/index.ts (hunk 18 above)
Comment (naripok) [reply types/index.ts:110]: Agree!

File: types/index.ts (hunk 17 above)
Comment (naripok) [reply types/index.ts:111]: Agree, removed.

File: src/engine/filters/index.ts (hunk 13 above)
Comment (naripok) [reply src/engine/filters/index.ts:27]: I think the intersection / vector distances are not actually computed once the type guard mismatch.
In the case where the VM optimizes the early returning on the checks and over the disjunction (some), it should be actually equivalent to the ternary type checking.
But I can be wrong... and I'm also not sure that it optimizes the disjunction...
I'll write a benchmark later to check if it blocks the client for too much time.

File: types/index.ts [hunk 20]
```diff
@@ -140,10 +127,13 @@ export interface EngineConditionRule {
   };
 }
 
-export interface EngineCondition {
+export type EngineConditionQuery<T extends AudienceDefinitionFilter> =
+  Pick<AudienceQueryDefinition, "featureVersion" | "queryProperty"> & T
```
Comment (ydennisy) [inline types/index.ts:131]: single quote :)

**Review by ydennisy — APPROVED** [review]
